jpa/eclipselink/src/main/java/org/jipijapa/eclipselink/JBossAS7TransactionController.java

```diff
@@ -0,0 +1,120 @@
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jipijapa.eclipselink;
+
+import javax.naming.NamingException;
+import javax.transaction.Synchronization;
+import javax.transaction.TransactionManager;
+import javax.transaction.TransactionSynchronizationRegistry;
+
+import org.eclipse.persistence.transaction.AbstractSynchronizationListener;
+import org.eclipse.persistence.transaction.jboss.JBossTransactionController;
+import org.jboss.logging.Logger;
+
+/**
+ * The transaction controller overrides some of the default behavior of eclipse
+ * link under the classes
+ * {@link org.eclipse.persistence.transaction.AbstractTransactionController} and
+ * {@link org.eclipse.persistence.transaction.JTATransactionController} to
+ * ensure that the unit of work synchronization is not taking place at too-late
+ * point in time.
+ *
+ * <P>
+ * REFERENCES: <br>
+ *
+ * <a href="https://issues.jboss.org/browse/WFLY-8954">Wildfly 10 with
+ * eclipselink Onscucess observer gets stale entity</a>
+ *
+ */
+public class JBossAS7TransactionController extends JBossTransactionController {
+
+    private static final Logger LOGGER = Logger.getLogger(JBossAS7TransactionController.class);
+
+    @Override
+    protected void registerSynchronization_impl(AbstractSynchronizationListener listener, Object txn) throws Exception {
+        try {
+            // (a) Our first approach is to register the listener on the
+            // transaction registry
+            // to ensure that this listener during the oncomplete phase of a
+            // transaction the weld CDI listener
+            registerSynchronizationUsingTransactionRegistry((Synchronization) listener);
+        } catch (Exception ingnoreE) {
+            // (b) Make sure that this error is made visible - we want the
+            // default approach to work 100% of the time - not just sometimes.
+            // We know that the fallback approach will yield us the bug
+            // See, WFLY-8954
+            // eclipselink will register the listener on the Transaction instead
+            // of on the TransactionRegistry
+            String errMsg = String.format(
+                    "Unexpected error took place while attempting to register the transaction listener: %1$s "
```

Comment:
We will need to move the logged warning/error messages to a resource file.  An example is org.jipijapa.JipiLogger, which extends org.jboss.logging.BasicLogger.  You could make this change, after the other changes and we explain more about adding/using a BasicLogger.

Reply:
Or maybe you could try adding the warning/error logging to the org.jipijapa.eclipselink.JBossLogger class, but we will still need to ask @jamezp to assign us an identifer range to use for the EclipseLink integration messages.

Reply:
IMO you can use whatever as long as it makes sense.